compiler/rustc_passes/src/region.rs

```diff
@@ -203,6 +204,116 @@ fn resolve_stmt<'tcx>(visitor: &mut RegionResolutionVisitor<'tcx>, stmt: &'tcx h
     visitor.cx.parent = prev_parent;
 }
 
+fn mark_local_terminating_scopes<'tcx>(expr: &'tcx hir::Expr<'tcx>) -> FxHashSet<hir::ItemLocalId> {
+    struct LocalAccessResolutionVisitor<'a> {
+        locals: &'a mut FxHashSet<hir::ItemLocalId>,
+    }
+    impl<'a> LocalAccessResolutionVisitor<'a> {
+        fn probe<'b>(&mut self, expr: &'b Expr<'b>) {
+            if self.locals.contains(&expr.hir_id.local_id) {
+                return;
+            }
+            let mut nested_expr = expr;
+            let mut ops = SmallVec::<[_; 4]>::new();
+            enum OpTy {
+                Ref,
+                Deref,
+                Project,
+                Local,
+            }
+            // Here we are looking for a chain of access to extract a place reference for
+            // a local variable.
+            // For instance, given `&*(&*x[y]).z().u`, `probe` will pick out
+            // the sub-expressions `&*x[y]` and `y` as expressions making accesses to locals.
+            // The place references generated can be discarded earlier as MIR allows,
+            // and, thus, can be assigned a smaller region.
+            loop {
+                match nested_expr.kind {
+                    hir::ExprKind::Path(hir::QPath::Resolved(
+                        _,
+                        hir::Path { res: hir::def::Res::Local(_), .. },
+                    )) => {
+                        ops.push((nested_expr, OpTy::Local));
+                        break;
+                    }
+                    hir::ExprKind::AddrOf(_, _, subexpr) => {
+                        ops.push((nested_expr, OpTy::Ref));
+                        nested_expr = subexpr;
+                    }
+                    hir::ExprKind::Unary(hir::UnOp::Deref, subexpr) => {
+                        ops.push((nested_expr, OpTy::Deref));
+                        nested_expr = subexpr;
+                    }
+                    hir::ExprKind::Field(subexpr, _) => {
+                        ops.push((nested_expr, OpTy::Project));
+                        nested_expr = subexpr;
+                    }
+                    hir::ExprKind::Index(subexpr, idxexpr) => {
+                        ops.push((nested_expr, OpTy::Project));
+                        nested_expr = subexpr;
+                        intravisit::walk_expr(self, idxexpr);
+                    }
+                    _ => {
+                        drop(ops);
+                        intravisit::walk_expr(self, expr);
+                        return;
+                    }
+                }
+            }
+            let mut ref_level = SmallVec::<[_; 4]>::new();
+            let mut ops_iter = ops.into_iter().rev();
+            ops_iter.next().unwrap();
```

Comment:
It might be worth a comment here saying that `ops_iter` always has at least one element because all of the arms from the previous `match` either push an item or return.

Reply:
I simplified the tracking a little bit, so that it does not need to track the local variable itself.

Reply:
Ah, even better. Thanks!